src/raft.rs

```diff
@@ -484,19 +484,56 @@ impl<T: Storage> Raft<T> {
         m.set_entries(RepeatedField::from_vec(ents));
         m.set_commit(self.raft_log.committed);
         if !m.get_entries().is_empty() {
-            match pr.state {
-                ProgressState::Replicate => {
-                    let last = m.get_entries().last().unwrap().get_index();
-                    pr.optimistic_update(last);
-                    pr.ins.add(last);
+            let last = m.get_entries().last().unwrap().get_index();
+            self.update_progress_state(last, pr);
+        }
+    }
+
+    fn update_progress_state(&mut self, last: u64, pr: &mut Progress) {
+        match pr.state {
+            ProgressState::Replicate => {
+                pr.optimistic_update(last);
+                pr.ins.add(last);
+            }
+            ProgressState::Probe => pr.pause(),
+            _ => panic!(
+                "{} is sending append in unhandled state {:?}",
+                self.tag, pr.state
+            ),
+        }
+    }
+
+    fn try_batching(
+        &mut self,
+        to: u64,
+        pr: &mut Progress,
+        ents: &mut Vec<Entry>,
+        term: u64,
+    ) -> (bool, bool, u64) {
+        // if MsgAppend for the reciver already exists, try_batching
+        // will append the entries to the existing MsgAppend
+        for msg in &mut self.msgs {
+            if msg.get_log_term() == term
+                && msg.get_msg_type() == MessageType::MsgAppend
+                && msg.get_to() == to
+            {
+                msg.set_index(pr.next_idx - 1);
+                msg.set_commit(self.raft_log.committed);
+                let mut batched_entries = msg.take_entries().into_vec();
```

Comment:
I worry that `into_vec` and following converting vector to RepeatedField again may reduce the performance. 

Maybe we need to introduce https://github.com/stepancheg/rust-protobuf/issues/280 

/cc @nrc

Reply:
The conversion from Vec to RepeatedField doesn't allocate or copy the Vec, it just moves the Vec into the RepeatedField, so will be cheap. It's not clear to me what was happening before, but there shouldn't be a performance hit unless we previously did not copy the message's entries at all

Reply:
Got it